ns for fx: add support for subgraph matching

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #52402 ns for fx: update graph matching to not match nodes with equal types
* #52395 ns for fx: support linear_relu for weight matching
* #52368 ns for fx: allow graph matching of parents of cat
* #52357 ns for fx: make unshadowed activation comparison work for N models
* #52356 ns for fx: make weights comparison work on N models
* **#52130 ns for fx: add support for subgraph matching**
* #52092 NS for FX: add test for a simple sparsenn model

Summary:

We have patterns like (F.linear, F.relu) which need to match
to (toq.linear_relu).  So, we need to match subgraphs.

This PR does the following:
* defines a "subgraph" as (start_node, end_node). The current assumption
is that subgraphs are simple, there is always a path from start_node to
end_node, and we can ignore any non-input args/kwargs of these nodes
for the purposes of matching and copying things. An example one node
subgraph is (F.linear, F.linear).  An example two node subgraph
is (F.linear, F.relu).
* changes the matching logic to iterate over subgraphs instead of nodes
* changes the NS core APIs to use subgraph pairs instead of node pairs:
1. for weights, we match on the start node
2. for unshadowed activations, we observe the end nodes
3. for shadowed activations, we copy the subgraph of a to graph c

TODO(before review) write up better, not ready for review yet

Test Plan:

```
python test/test_quantization.py TestFXNumericSuiteCoreAPIs
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26403092](https://our.internmc.facebook.com/intern/diff/D26403092)

vkuzo committed:
[wip] ns for fx: add support for subgraph matching

Summary:

We have patterns like (F.linear, F.relu) which need to match
to (toq.linear_relu).  So, we need to match subgraphs.

This PR does the following:
* defines a "subgraph" as (start_node, end_node). The current assumption
is that subgraphs are simple, there is always a path from start_node to
end_node, and we can ignore any non-input args/kwargs of these nodes
for the purposes of matching and copying things. An example one node
subgraph is (F.linear, F.linear).  An example two node subgraph
is (F.linear, F.relu).
* changes the matching logic to iterate over subgraphs instead of nodes
* changes the NS core APIs to use subgraph pairs instead of node pairs:
1. for weights, we match on the start node
2. for unshadowed activations, we observe the end nodes
3. for shadowed activations, we copy the subgraph of a to graph c

TODO(before review) write up better, not ready for review yet

Test Plan:

TODO before land: better test plan

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
Update on "[wip] ns for fx: add support for subgraph matching"


Summary:

We have patterns like (F.linear, F.relu) which need to match
to (toq.linear_relu).  So, we need to match subgraphs.

This PR does the following:
* defines a "subgraph" as (start_node, end_node). The current assumption
is that subgraphs are simple, there is always a path from start_node to
end_node, and we can ignore any non-input args/kwargs of these nodes
for the purposes of matching and copying things. An example one node
subgraph is (F.linear, F.linear).  An example two node subgraph
is (F.linear, F.relu).
* changes the matching logic to iterate over subgraphs instead of nodes
* changes the NS core APIs to use subgraph pairs instead of node pairs:
1. for weights, we match on the start node
2. for unshadowed activations, we observe the end nodes
3. for shadowed activations, we copy the subgraph of a to graph c

TODO(before review) write up better, not ready for review yet

Test Plan:

TODO before land: better test plan

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
Update on "[wip] ns for fx: add support for subgraph matching"


Summary:

We have patterns like (F.linear, F.relu) which need to match
to (toq.linear_relu).  So, we need to match subgraphs.

This PR does the following:
* defines a "subgraph" as (start_node, end_node). The current assumption
is that subgraphs are simple, there is always a path from start_node to
end_node, and we can ignore any non-input args/kwargs of these nodes
for the purposes of matching and copying things. An example one node
subgraph is (F.linear, F.linear).  An example two node subgraph
is (F.linear, F.relu).
* changes the matching logic to iterate over subgraphs instead of nodes
* changes the NS core APIs to use subgraph pairs instead of node pairs:
1. for weights, we match on the start node
2. for unshadowed activations, we observe the end nodes
3. for shadowed activations, we copy the subgraph of a to graph c

TODO(before review) write up better, not ready for review yet

Test Plan:

TODO before land: better test plan

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
Update on "[wip] ns for fx: add support for subgraph matching"


Summary:

We have patterns like (F.linear, F.relu) which need to match
to (toq.linear_relu).  So, we need to match subgraphs.

This PR does the following:
* defines a "subgraph" as (start_node, end_node). The current assumption
is that subgraphs are simple, there is always a path from start_node to
end_node, and we can ignore any non-input args/kwargs of these nodes
for the purposes of matching and copying things. An example one node
subgraph is (F.linear, F.linear).  An example two node subgraph
is (F.linear, F.relu).
* changes the matching logic to iterate over subgraphs instead of nodes
* changes the NS core APIs to use subgraph pairs instead of node pairs:
1. for weights, we match on the start node
2. for unshadowed activations, we observe the end nodes
3. for shadowed activations, we copy the subgraph of a to graph c

TODO(before review) write up better, not ready for review yet

Test Plan:

TODO before land: better test plan

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26403092](https://our.internmc.facebook.com/intern/diff/D26403092)

[ghstack-poisoned]

vkuzo committed:
Update on "[wip] ns for fx: add support for subgraph matching"


Summary:

We have patterns like (F.linear, F.relu) which need to match
to (toq.linear_relu).  So, we need to match subgraphs.

This PR does the following:
* defines a "subgraph" as (start_node, end_node). The current assumption
is that subgraphs are simple, there is always a path from start_node to
end_node, and we can ignore any non-input args/kwargs of these nodes
for the purposes of matching and copying things. An example one node
subgraph is (F.linear, F.linear).  An example two node subgraph
is (F.linear, F.relu).
* changes the matching logic to iterate over subgraphs instead of nodes
* changes the NS core APIs to use subgraph pairs instead of node pairs:
1. for weights, we match on the start node
2. for unshadowed activations, we observe the end nodes
3. for shadowed activations, we copy the subgraph of a to graph c

TODO(before review) write up better, not ready for review yet

Test Plan:

TODO before land: better test plan

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26403092](https://our.internmc.facebook.com/intern/diff/D26403092)

[ghstack-poisoned]

raghuramank100 commented on torch/quantization/ns/graph_matcher.py:
```diff
@@ -87,12 +88,47 @@ def get_non_matchable_modules() -> Set[Callable]:
+        (F.relu, F.linear),
+    ])
+
+# TODO(future PR): we should see if we can reuse quantization's fusion
```
Yes, this is a good idea. In general, would the entire graph_matcher logic be shared with fx quantization passes?

vkuzo replied:
that would be a good long term state.  I'd imagine most of the benefits to come from sharing the fusion patterns, the matching logic itself may not generalize.

raghuramank100 commented on torch/quantization/ns/graph_matcher.py:
```diff
@@ -111,27 +147,52 @@ def __init__(
+            # if we match linear-relu backwards, we will always skip the
+            # relu node and attempt to match the linear node.  This can
+            # be made configurable later if needed.
+            for _reverse_fusion_ops in get_reversed_fusions():
```
Do we always pick the largest fusion here?

vkuzo replied:
Ideally the ordering in the default would be followed, with the ability for user to customize.  It doesn't handle that yet.  We'd have to add that logic as soon as we have more than one fusion.

raghuramank100 commented on torch/quantization/ns/graph_matcher.py:
```diff
@@ -185,16 +246,18 @@ def _node_a_related_to_b(
     # for now, use node name.
     # TODO(future PR): find a better solution
-    return node_b.name
+    return end_node_b.name
```
Should this be start_node_b.name+end_node_b.name so that it is clear that it is a sub-graph and not a single node?

vkuzo replied:
Sounds good to me.  In general naming of nodes is something we should talk through in detail for all of these APIs, saving that for a future PR.  Thankfully changing that later should be low eng cost, so it doesn't have to be perfect now.

raghuramank100 commented on torch/quantization/ns/graph_matcher.py:
```diff
@@ -87,12 +88,47 @@ def get_non_matchable_modules() -> Set[Callable]:
     and continuing backwards.
-    1. Returns matchable nodes, in order
-    2. Skips over non-matchable nodes
+    1. Returns matchable subgraphs, in order. A subgraph is defined by
```
Sorry, dont follow this logic: Where is the sub-graph specified in the init? I dont see the end_node being defined.

vkuzo replied:
this is in the `__next__` function.  We pop a potential end node off the stack.  If the end node is matchable, we try to peek backwards to check for a potential fusion pattern.  We return (end_node, end_node) if there is no fusion pattern, or (start_node, end_node) if there is a fusion pattern ending at end_node.

vkuzo committed:
Update on "[wip] ns for fx: add support for subgraph matching"


Summary:

We have patterns like (F.linear, F.relu) which need to match
to (toq.linear_relu).  So, we need to match subgraphs.

This PR does the following:
* defines a "subgraph" as (start_node, end_node). The current assumption
is that subgraphs are simple, there is always a path from start_node to
end_node, and we can ignore any non-input args/kwargs of these nodes
for the purposes of matching and copying things. An example one node
subgraph is (F.linear, F.linear).  An example two node subgraph
is (F.linear, F.relu).
* changes the matching logic to iterate over subgraphs instead of nodes
* changes the NS core APIs to use subgraph pairs instead of node pairs:
1. for weights, we match on the start node
2. for unshadowed activations, we observe the end nodes
3. for shadowed activations, we copy the subgraph of a to graph c

TODO(before review) write up better, not ready for review yet

Test Plan:

TODO before land: better test plan

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26403092](https://our.internmc.facebook.com/intern/diff/D26403092)

[ghstack-poisoned]

vkuzo committed:
rebase on "[wip] ns for fx: add support for subgraph matching"


Summary:

We have patterns like (F.linear, F.relu) which need to match
to (toq.linear_relu).  So, we need to match subgraphs.

This PR does the following:
* defines a "subgraph" as (start_node, end_node). The current assumption
is that subgraphs are simple, there is always a path from start_node to
end_node, and we can ignore any non-input args/kwargs of these nodes
for the purposes of matching and copying things. An example one node
subgraph is (F.linear, F.linear).  An example two node subgraph
is (F.linear, F.relu).
* changes the matching logic to iterate over subgraphs instead of nodes
* changes the NS core APIs to use subgraph pairs instead of node pairs:
1. for weights, we match on the start node
2. for unshadowed activations, we observe the end nodes
3. for shadowed activations, we copy the subgraph of a to graph c

TODO(before review) write up better, not ready for review yet

Test Plan:

TODO before land: better test plan

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26403092](https://our.internmc.facebook.com/intern/diff/D26403092)

[ghstack-poisoned]

jerryzh168 commented on torch/quantization/ns/graph_matcher.py:
```diff
@@ -87,12 +88,47 @@ def get_non_matchable_modules() -> Set[Callable]:
     ])
 
-class _NSGraphMatchableNodesIterator:
+def get_reversed_fusions() -> Set[Tuple[Callable, Callable]]:
```
is this only going to support matching 2 nodes?

vkuzo replied:
the main restriction is that the subgraph needs one start point and one end point.  Support for a length of 3 nodes can be added here with low eng cost.  This type would become `Union[Tuple[Callable, Callable], Tuple[Callable, Callable, Callable]]`.

note: currently this is not a list because lists are not hashable.  In general, I'm flexible on the exact data structure, happy to revise it if there is a better solution.

hx89 commented on torch/quantization/ns/graph_matcher.py:
```diff
@@ -224,73 +296,99 @@ def get_matching_node_pairs(
+            # note: relatedness is checked on the start node, i.e.
+            # if a linear-relu pattern is checked, we would check for relatedness
+            # of the linear
+            if not _node_a_related_to_b(cur_start_node_a, cur_start_node_b,
```
Since subgraph matching is introduced to graph matcher, would it be natural to extend the relatedness checking to support subgraph also, which might be more general?

vkuzo replied:
Sure, that could make sense.  Is there a use case in mind?

Since this is a private API, the eng cost of changing this in the future would be low.

hx89 replied:
Yeah we can think about it in later PR. I don't have specific use case in mind, just thinking if we can relate for example the subgraph (linear, relu) to linear_relu directly, then it would be self-explanatory and we may not need to note "relatedness is checked on the start node".

vkuzo commented on torch/quantization/ns/graph_matcher.py:
```diff
@@ -111,27 +147,52 @@ def __init__(
                 continue
-            self.seen_nodes.add(cur_node)
+
+            # for subgraphs which are single nodes, start_node == end_node
```
sure, the node here is the actual node object (not node type or name, etc).  So, two distinct F.linear nodes can be used as a start point and an endpoint.  For example,

```
linear1 -> linear2 -> linear3

# a subgraph of just linear1
(linear1, linear1)

# a subgraph of 1..3
(linear1, linear3)
```

vkuzo commented on torch/quantization/ns/graph_matcher.py:
```diff
@@ -185,16 +246,18 @@ def _node_a_related_to_b(
-    if node_b.op == 'call_module':
-        assert isinstance(node_b.target, str)
-        return node_b.target
+    if end_node_b.op == 'call_module':
```
sounds reasonable to me, happy to update that in a separate PR.  In general this part is pretty self-contained, so we can make updates to this easily.

raghuramank100 commented on torch/quantization/ns/graph_passes.py:
```diff
@@ -266,9 +311,12 @@ def create_a_shadows_b(
-    nodes_to_instrument_b_to_a = {}
-    for match_name, (node_a, node_b) in matched_node_pairs.items():
-        nodes_to_instrument_b_to_a[node_b] = node_a
+    node_b_to_matched_subgraph_a = {}
```
It is quite useful to insert shadow at multiple levels.  Assuming in a later PR we will support shadowing a sub-graph instead of a single node.

vkuzo replied:
sounds reasonable

raghuramank100 commented on torch/quantization/ns/graph_passes.py:
```diff
@@ -330,7 +379,7 @@ def load_arg(a):
             # subgraph so far:
             #
-            #       dtype_cast_node --> node_a_copy
+            #       dtype_cast_node --> subgraph_a_copy
```
Should the cast be at the input of node_c?

vkuzo replied:
correct.  The code does this, but the comment is wrong, thanks for catching.  Will update before landing.

vkuzo committed:
rebase on "ns for fx: add support for subgraph matching"


Summary:

We have patterns like (F.linear, F.relu) which need to match
to (toq.linear_relu).  So, we need to match subgraphs.

This PR does the following:
* defines a "subgraph" as (start_node, end_node). The current assumption
is that subgraphs are simple, there is always a path from start_node to
end_node, and we can ignore any non-input args/kwargs of these nodes
for the purposes of matching and copying things. An example one node
subgraph is (F.linear, F.linear).  An example two node subgraph
is (F.linear, F.relu).
* changes the matching logic to iterate over subgraphs instead of nodes
* changes the NS core APIs to use subgraph pairs instead of node pairs:
1. for weights, we match on the start node
2. for unshadowed activations, we observe the end nodes
3. for shadowed activations, we copy the subgraph of a to graph c

TODO(before review) write up better, not ready for review yet

Test Plan:

```
python test/test_quantization.py TestFXNumericSuiteCoreAPIs
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26403092](https://our.internmc.facebook.com/intern/diff/D26403092)

[ghstack-poisoned]

vkuzo committed:
fix asan on "ns for fx: add support for subgraph matching"


Summary:

We have patterns like (F.linear, F.relu) which need to match
to (toq.linear_relu).  So, we need to match subgraphs.

This PR does the following:
* defines a "subgraph" as (start_node, end_node). The current assumption
is that subgraphs are simple, there is always a path from start_node to
end_node, and we can ignore any non-input args/kwargs of these nodes
for the purposes of matching and copying things. An example one node
subgraph is (F.linear, F.linear).  An example two node subgraph
is (F.linear, F.relu).
* changes the matching logic to iterate over subgraphs instead of nodes
* changes the NS core APIs to use subgraph pairs instead of node pairs:
1. for weights, we match on the start node
2. for unshadowed activations, we observe the end nodes
3. for shadowed activations, we copy the subgraph of a to graph c

TODO(before review) write up better, not ready for review yet

Test Plan:

```
python test/test_quantization.py TestFXNumericSuiteCoreAPIs
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26403092](https://our.internmc.facebook.com/intern/diff/D26403092)

[ghstack-poisoned]
